### PR TITLE
fix(filename_template): `[hash:6]` in `entryFileNames` does not get replaced with hash

### LINKED ISSUE #2206: [Bug]: `[hash:6]` in `entryFileNames` does not get replaced with hash
### Reproduction link or steps

1. Open https://stackblitz.com/edit/github-6jajke-nva34g?file=rolldown.config.mjs
2. `npm run build` runs automatically
3. see `dist/entry-[hash:6].js` is generated
4. run `npm run build:up` to run rollup
5. see `dist/entry-q8VKMz.js` is generated

### What is expected?

`dist/entry-hash.js` (`hash` part can be any 6-character long hash) is generated

### What is actually happening?

`dist/entry-[hash:6].js` is generated (`[hash:6]` is not replaced with the real hash)

### System Info

```Shell
Stackblitz

  System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.20.3 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.15.6 - /usr/local/bin/pnpm
  npmPackages:
    rolldown: https://pkg.pr.new/rolldown@d6233d8 => 0.13.2
```

### Any additional comments?

This is not something urgent.

### PR DESCRIPTION
**Pull request by shulaoda**
### Description

closes #2206

### CODE REVIEW

File: crates/rolldown_common/src/inner_bundler_options/types/filename_template.rs [hunk 1]
```diff
@@ -26,14 +29,27 @@ pub struct FileNameRenderOptions<'me> {
   pub ext: Option<&'me str>,
 }
 
+static HASH_PLACEHOLDER: LazyLock<Regex> =
+  LazyLock::new(|| Regex::new(r"\[hash(:(\d*))?]").expect("Invalid hash regex"));
```
Comment (Boshen) [inline crates/rolldown_common/src/inner_bundler_options/types/filename_template.rs]: Similar to aa347f5, regex is unnecessary here.

**Comment by IWANABETHATGUY** [comment]
only one test in js side is enough, since what you need to ensure is file name has `hash` in dist

File: crates/rolldown_common/src/inner_bundler_options/types/filename_template.rs [hunk 2]
```diff
@@ -33,7 +34,7 @@ impl FilenameTemplate {
       tmp = tmp.replace("[name]", name);
     }
     if let Some(hash) = options.hash {
-      tmp = tmp.replace("[hash]", hash);
+      tmp = tmp.replace(options.hash_pattern.unwrap_or("[hash]"), hash);
```
Comment (7086cmd) [inline crates/rolldown_common/src/inner_bundler_options/types/filename_template.rs]: I recommend completing the task in the file_emitter.rs file. For reference, the relevant code can be found [here](https://github.com/rolldown/rolldown/blob/main/crates/rolldown_common/src/file_emitter.rs#L86).

File: packages/rolldown/tests/fixtures/output/fileNames/hash/_config.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import type { RolldownOutputChunk } from 'rolldown'
+import { defineTest } from '@tests'
+import { expect } from 'vitest'
+
+export default defineTest({
+  config: {
+    output: {
+      entryFileNames: '[name]-[hash:6].js',
+      chunkFileNames: '[name]-[hash:7].js',
```
Comment (7086cmd) [inline packages/rolldown/tests/fixtures/output/fileNames/hash/_config.ts:9]: What if we need more than 8?

**Review by 7086cmd — CHANGES_REQUESTED** [review]

File: packages/rolldown/tests/fixtures/output/fileNames/hash/_config.ts (hunk 3 above)
Comment (shulaoda) [reply packages/rolldown/tests/fixtures/output/fileNames/hash/_config.ts:9]: We seem to support hash lengths of `5-22`.
https://github.com/rolldown/rolldown/blob/e52ef2f22015b919e958bdcf332bfca7425144d2/crates/rolldown/src/utils/hash_placeholder.rs#L44

**Comment by shulaoda** [comment]
I have a question: 
Is it possible for there to be a situation where `[hash]-[hash:6].js` results in duplicate hashes? If so, there might be some errors with our current implementation.

File: packages/rolldown/tests/fixtures/output/fileNames/hash/_config.ts (hunk 3 above)
Comment (7086cmd) [reply packages/rolldown/tests/fixtures/output/fileNames/hash/_config.ts:9]: Is [this](https://github.com/rolldown/rolldown/blob/main/crates/rolldown_common/src/file_emitter.rs#L86) `8` related?

**Comment by shulaoda** [comment]
Both `esbuild` and `rollup` handle duplicate `[hash]-[hash]`, perhaps we should also support this feature.

File: packages/rolldown/tests/fixtures/output/fileNames/hash/_config.ts (hunk 3 above)
Comment (shulaoda) [reply packages/rolldown/tests/fixtures/output/fileNames/hash/_config.ts:9]: Got it, I will improve.

**Comment by IWANABETHATGUY** [comment]
> Both `esbuild` and `rollup` handle duplicate `[hash]-[hash]`, perhaps we should also support this feature.

You could file another issue to track this.

File: crates/rolldown/src/utils/hash_placeholder.rs [hunk 4]
```diff
@@ -22,16 +22,21 @@ const CHARS: &[u8] = b"0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTU
...
-    out_string.push(CHARS[current_digit as usize] as char);
+    index += 1;
+
     if value == 0 {
       break;
     }
   }
-  out_string
+
+  std::str::from_utf8(&buffer[..index]).unwrap().to_string()
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/utils/hash_placeholder.rs]: String::from_utf8_lossy(&a).to_string()

File: crates/rolldown/src/utils/hash_placeholder.rs [hunk 5]
```diff
@@ -42,20 +47,20 @@ pub struct HashPlaceholderGenerator {
...
-        padding.push('0');
-      }
-      seed_base64 = [padding, seed_base64].concat();
-    }
+    let seed_base64 = to_base64(self.seed);
+
+    // TODO(hyf0): improve this
+    assert!(seed_base64.len() <= allow_middle_len, "seed is too large");
+
+    let mut placeholder = String::with_capacity(len + 5);
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/utils/hash_placeholder.rs]: pre alloc should consider all sub string length.

**Review by IWANABETHATGUY — APPROVED** [review]

**Pull request merged**